netsync/manager.go

```diff
@@ -427,24 +431,38 @@ func (sm *SyncManager) current() bool {
 // restartSyncIfNeeded finds a new sync candidate if we're not expecting any
 // blocks from the current one.
 func (sm *SyncManager) restartSyncIfNeeded() {
-	if sm.syncPeer != nil {
-		syncPeerState, exists := sm.peerStates[sm.syncPeer]
-		if exists {
-			isWaitingForBlocks := func() bool {
-				syncPeerState.requestQueueMtx.Lock()
-				defer syncPeerState.requestQueueMtx.Unlock()
-				return len(syncPeerState.requestedBlocks) != 0 || len(syncPeerState.requestQueues[wire.InvTypeSyncBlock].queue) != 0
-			}()
-			if isWaitingForBlocks {
-				return
-			}
-		}
+	sm.syncPeerLock.Lock()
+	defer sm.syncPeerLock.Unlock()
+	shouldReplaceSyncPeer, err := sm.shouldReplaceSyncPeer()
+	if err != nil {
+		panic(err)
+	}
+
+	if !shouldReplaceSyncPeer {
+		return
 	}
 
 	sm.syncPeer = nil
 	sm.startSync()
 }
 
+func (sm *SyncManager) shouldReplaceSyncPeer() (bool, error) {
+	if sm.syncPeer == nil {
+		return true, nil
+	}
+
+	syncPeerState, exists := sm.peerStates[sm.syncPeer]
```

Comment:
Actually, in case we panic anyway, why not let sm.peerStates[sm.syncPeer] panic to start with?
I mean, just do `syncPeerState := sm.peerStates[sm.syncPeer]`.

If you assume this never happens anyway, then you are saving some code complexity, and if you think this should be handled properly, then handle it properly, not with a panic